module4/Homework4Playground.playground/Contents.swift

```diff
@@ -0,0 +1,211 @@
+
+//- Create a function called evenNumbersArray that takes a parameter of [Int] (array of Int) and returns [Int]. The array of Int returned should contain all the even numbers in the array passed. Call the function passing the nums array and print the output.
+
+func evenNumbersArray(_ array: [Int]) -> [Int] {
+  array.filter({ $0 % 2 == 0})
+}
+
+let evenNums = evenNumbersArray(nums)
+print("evenNums \(evenNums)")
+
+//- Create a function called sumOfArray that takes a parameter of [Int?] and returns an Int. The function should return the sum of the array values passed that are not nil. Call the function passing the numsWithNil array, and print out the results.
+
+func sumOfArray(_ array: [Int?]) -> Int {
+  array.compactMap({ $0 }).reduce(0, { $0 + $1 })
+}
+let nonNilSum = sumOfArray(numsWithNil)
+print("nonNilSum \(nonNilSum)")
+
+//- Create a function called commonElementsSet that takes two parameters of [Int] and returns a Set<Int> (set of Int.) The function will return a Set<Int> of the values in both arrays.
+
+func commonElementsSet(_ array1: [Int], _ array2: [Int]) -> Set<Int> {
+  return Set(array2).intersection(array1)
+}
+
+//Call the function commonElementsSet passing the arrays numsBy2, numsBy4, and print out the results.
+
+let commonBy2AndBy4 = commonElementsSet(numsBy2, numsBy4).sorted()
+print("commonBy2AndBy4 \(commonBy2AndBy4)")
+
+//d) Create a struct called Square that has a stored property called sideLength and a computed property called area. Create an instance of Square and print out the area.
+
+print("d) ---------------")
+
+struct Square {
+  let sideLength:  Double
+  var area: Double {
+    sideLength * sideLength
+  }
+}
+
+let sq = Square(sideLength: 4)
+print("sq.area \(sq.area)")
+
+//Part 3 - Above and Beyond
+
+//Create a protocol called Shape with a calculateArea() -> Double method. Create two structs called Circle and Rectangle that conform to the protocol Shape. Both Circle and Rectangle should have appropriate stored properties for calculating the area.
+
+protocol Shape {
+  func calculateArea() -> Double
+}
+
+struct Circle: Shape {
+  let radius: Double
+  
+  func calculateArea() -> Double {
+    Double.pi * radius * radius
+  }
+}
+
+struct Rectangle: Shape {
+  let width: Double
+  let height: Double
+
+  func calculateArea() -> Double {
+    width * height
+  }
+}
+
+//Create instances of Circle and Rectangle and print out the area for each.
+
+let circle = Circle(radius: 3)
+print("circle area: \(circle.calculateArea())")
+
+let rect = Rectangle(width: 3, height: 4)
+print("rect area: \(rect.calculateArea())")
+
+//Next, extend the protocol Shape to add a new method called calculateVolume() -> Double.
+
+
+// Interesting - it seems you can extend a protocol w/out defining a default implementation. I guess that makes sense?
```

Comment:
? You do provide a default implementation for your new func. I don't understand your comment.

Reply:
typo: 
I meant:
"it seems you * can't * extend a protocol w/out defining a default implementation."

Which was surprising to me when I wrote this. But it makes sense in that, if others have already implemented the protocol, then you cant add a function that they then must go implement to conform.